### DISCUSSION TITLE
Vim mode / open current file in external editor

### DISCUSSION THREAD

**Discussion by idealemu** [vkbo/novelWriter#2027] [Ideas]
Hi,

How hard would it be to implement a "vim mode" for basic movements in the editor ? I know quite a few open source projects have this feature, or a plugin that enables it, we could look at for inspiration. Since the idea is mainly relevant for editing, IMHO, perhaps this could also be achieved by 'just' having a feature to open the current file in an external editor ? 

I already do this by going into the filesystem and opening the corresponding file if I have specific changes I know are easier to make with vim, but since the files are named for the database it is a slightly cumbersome workflow at the moment.

**Comment by vkbo** [comment]
Well, you're not supposed to edit files externally as it causes issues with the index and internal meta data. If you do so, it's at your own risk, and it's not something the editor is going to support.

As for vim mode, I guess it could be done. Not sure how it would interfere with other functionality and native Qt features. For the time being, I'm focusing my efforts on editor features, for which the backlog is still pretty big.

**Comment by Polygon1556** [comment]
I second the VIM editing mode in the main text editor (not opening in an external program). I wish I had the programming skills to contribute this feature. The "vibe" of NovelWriter is already pretty clearly from a coding or programming background (if that makes sense), and I think a lot of people would love a VIM mode.

I use NeoVim for a lot of my personal writing, but for novel writing I need the NovelWriter organization and tagging features, so adding a VIM mode would make it perfect.

**Reply by vkbo** [reply]
Yeah, it would need to be contributed because I'm not a vim user so don't really know what "vim mode" is, although I can guess. All I know about vim is Insert/Escape and :wq 😄 

But you're right about novelWriter. It definitely has a code editor vibe. It's deliberate, but I try to make it not the only way to use the app.

**Reply by idealemu** [reply]
Ok, I would be willing to try to make a very basic vim mode, I'll give it a try later today and report back ;)

**Reply by idealemu** [reply]
A vim mode for NovelWriter should probably focus on just the very basics:

- "h" "j" "k" "l" for movement (maybe basic more advanced vim motions such as "w" "e" "b")
- "dd" for line deletion, "x" for current character deletion
- "y" to yank (copy) (again maybe some basic combinations such as "yw" etc)
- ESC, "i", "a", "v" to move between vim modes
- "G" "gg" for navigation 
- "^," "$" go to start/end of line 
- "o" "O" open new line below/above
- "u" undo and Ctrl-r – redo

One challenge I can already foresee is grabbing the Escape key event to switch to vim's normal mode, as I believe the event is probably captured at a higher level and not passed on. 
If you were to provide insights into how each functionality could be implemented/used in the best way in NovelWriter I could integrate that with some key events and logic to make a basic vim mode :smile:

**Reply by vkbo** [reply]
It would need to go in the doceditor.py file.

https://github.com/vkbo/novelWriter/blob/21ce0304631e337af0ab4854c4ca860475e99df1/novelwriter/gui/doceditor.py#L932-L969

All key presses are already captured, as this is where auto complete and a few other features hook in, so I guess that's the natural place. I guess some flag that will block key presses from reaching the editor when in other than writing mode would be needed too. Key presses are forwarded to the editor itself when `super().keyPressEvent(event) ` is called, so all you have to do in the other modes is to block that. Movement and changes to the text can be done by creating a QTextCursor and passing it to the editor. There are plenty of examples in the existing code that does this.

**Reply by vkbo** [reply]
Actually, auto-complete is hooked into the document change handler, not the keyPressEvent, but that's besides the point. If you need to capture text changes, you need to hook in there instead or in addition.

https://github.com/vkbo/novelWriter/blob/21ce0304631e337af0ab4854c4ca860475e99df1/novelwriter/gui/doceditor.py#L1079-L1119

**Reply by idealemu** [reply]
> I guess some flag that will block key presses from reaching the editor when in other than writing mode would be needed too. Key presses are forwarded to the editor itself when `super().keyPressEvent(event) ` is called, so all you have to do in the other modes is to block that. 

I can get a basic normal mode vim motions to work by simply processing vim related keys before going to the rest of key events. IE: if in insert mode work as normal, but if in normal mode handle keys as vim keys and just return early from the key handler.

> Movement and changes to the text can be done by creating a QTextCursor and passing it to the editor. There are plenty of examples in the existing code that does this.

I think that part should be pretty straightforward indeed.

**Reply by idealemu** [reply]
> Actually, auto-complete is hooked into the document change handler, not the keyPressEvent, but that's besides the point. If you need to capture text changes, you need to hook in there instead or in addition.

So auto-complete grabs all keys ? Including Escape ? (I couldn't figure this out just by looking at the code snippet)

Online sources typically show that grabbing escape can be done by creating a [Qt shortcut bound to escape](https://stackoverflow.com/a/56895453) to grab the key event before anything else. 
If nothing else is grabbing it, or if it is still passed down then we can simply get it like any other key apparently like such `if event.key() == Qt.Key_Escape`.

**Reply by vkbo** [reply]
No, that is a function that is called whenever the content of the document changes. I was just mistaken in my first response when I said it intercepted the key press event itself.

The auto complete works by taking the last character you typed, look backwards to see if there is a combo that needs replacing, and then does an edit using the QTextCursor. So it's a good example of how to edit by cursor.

**Reply by vkbo** [reply]
> Online sources typically show that grabbing escape can be done by creating a [Qt shortcut bound to escape](https://stackoverflow.com/a/56895453) to grab the key event before anything else. If nothing else is grabbing it, or if it is still passed down then we can simply get it like any other key apparently like such `if event.key() == Qt.Key_Escape`.

novelWriter actually does this already. Escape is captured globally here:

https://github.com/vkbo/novelWriter/blob/21ce0304631e337af0ab4854c4ca860475e99df1/novelwriter/guimain.py#L308-L311

And processed here:

https://github.com/vkbo/novelWriter/blob/21ce0304631e337af0ab4854c4ca860475e99df1/novelwriter/guimain.py#L1326-L1333

You can also associate a shortcut with a specific widget, but that's a bit tricky with the editor because it is a combo of multiple widgets, some of which are free floating and sometimes hidden, hence the custom logic to just capture it globally and use the if statement to decide what to do. You probably need to hook in there to handle the Escape, or try to find a better way to solve that.

**Reply by vkbo** [reply]
In any case, I can help with a lot of these details if you otherwise have a proof of concept solution. Keep in mind that every line of code added also needs to be covered by unit tests in novelWriter.

**Reply by idealemu** [reply]
> novelWriter actually does this already. Escape is captured globally here

Super ! This is what I have been looking for, I just couldn't find where it was. If I just add a global vim mode state I can get a basic proof of concept to work I think !
And yes I have read through the docs and know about the tests of course.

**Reply by vkbo** [reply]
I created a feature ticket for this: #2493

Let's keep the implementation discussion going there.

**Reply by Polygon1556** [reply]
Thank you @vkbo and @idealemu so much! I can't believe this might actually happen!!!!

**Comment by SuperUserNameMan** [comment]
May i ask what a "vim mode" is ?

**Reply by idealemu** [reply]
Sure ! A "vim mode" is another editing mode that can be implented by an editor that emulates some of the functionalities of the [vim editor](https://www.vim.org/). [Vim is a text editor ](https://en.wikipedia.org/wiki/Vim_(text_editor) that is quite popular amongst (I would say) the linux crowd, or people who often work from the terminal, although it doesn't need to be.

The main 'point' of it is a text editor that is fully customizable through a configuration file (this isn't often included in vim modes in other editors), and where all actions can be taken with the keyboard. Mastery of vim, or just knowing the basics, provides a lot of speed, and usually once people like it they usually can't go without it, hence "vim modes" being added to a whole bunch of other editors ! :smile:

**Reply by SuperUserNameMan** [reply]
Thanks for clarifying what you were referring to by "vim mode". (i wasn't sure when i googled).

I'm more a nano/gedit guy. Will this VIM mode that you want to add to novelWritter be optional ?

**Reply by vkbo** [reply]
Yes, it will be optional 😃
